### ISSUE TITLE
Support Controller Input for Pause Menu 

### ISSUE THREAD

**Issue by Ltek** [syd711/vpin-studio#130]
Pause menu will not to show. Overlay menu shows properly, game is Pausing properly

<Log attached>

**To Reproduce**
Configure system as such (see 3 attached screenshots)
A. VPX settings: 'P' mapped to Pause; no button assigned  (stock configuration)
B. Joy2Key maps 'Z' to a cabinet button
C. VPS maps 'Z' to Pause Menu (screenshot attached)

1. Run Popper
2. Load Table
3. Press P : game pauses, no VPS menu shows
4. Press P again : game unpauses
5. Press Z : nothing happens

**Screenshots**
![image](https://github.com/syd711/vpin-studio/assets/1626802/ab835cd1-811c-45e3-beef-19947c418bf1)
![image](https://github.com/syd711/vpin-studio/assets/1626802/4cbf29fc-0da5-4acc-801b-8e3a8d4582b6)
![image](https://github.com/syd711/vpin-studio/assets/1626802/e157f428-3592-447f-8f72-f89ec58cd0b2)
[vpin-studio-server.log](https://github.com/syd711/vpin-studio/files/14106974/vpin-studio-server.log)

**Comment by syd711** [comment]
A perfect report, thanks for this! I'll look into it.

**Comment by syd711** [comment]
Can you verify that you have these 2x entries in the emulator setup?

![image](https://github.com/syd711/vpin-studio/assets/1169826/082bbc33-1723-4743-a783-62a6fd37d896)

They should be generated there automatically, but I got a similar issue recently where the entry was not there.

**Comment by syd711** [comment]
I just verified this. The server log shows the message:

>  Skipped showing start menu: no game status found.

The game status is set through these curl calls. So once you add them, the pause menu should appear.

**Comment by Ltek** [comment]
> I just verified this. The server log shows the message:
> 
> > Skipped showing start menu: no game status found.
> 
> The game status is set through these curl calls. So once you add them, the pause menu should appear.

they are there...

![image](https://github.com/syd711/vpin-studio/assets/1626802/f9047bc6-db1a-466c-93a9-f2d630f59a99)

**Comment by syd711** [comment]
> B. Joy2Key maps 'Z' to a cabinet button

This is the problem here: right now I only listen to native key events. So even if I read the "jobstick -> key" mapping from Popper I have to add additonal stuff to support joysticks. I already found libraries for this and keep you posted about the progress here.

**Comment by Ltek** [comment]
My understanding is all hardware drivers (real hardware keyboards, generic usb button converters, etc) and software converter apps (J2K, Popper, VPX, etc) convert commands to OS native Virtual-key codes, which are device-independent.

If VPS is listening for what the OS is getting (the virtual key code) then it will not matter what device or hardware sends the code. VPS will not know, or care so commands sent from J2K, popper, etc will all be heard by VPS. And the benefit is that any command can be heard by VPS and leveraged if so desired.

**Comment by syd711** [comment]
I did investigate and played around with some APIs a bit. With Java, the whole controller support is a bit tricky. There is a library available, but I wasn't able to achieve the expected results yet and the installation is tricky too. So this ticket may take a while longer.

**Comment by Ltek** [comment]
I dont know java, at all but found this...

https://www.educative.io/answers/how-to-listen-to-and-take-action-on-keyboard-strokes-in-java

https://stackoverflow.com/questions/2168968/can-java-see-activity-of-my-keyboard

https://stackoverflow.com/questions/22955770/how-to-find-key-listener-on-any-where-in-window

**Comment by syd711** [comment]
I changed the title so that it matches the overall problem.

**Comment by syd711** [comment]
If I can avoid it, I'd like to skip the controller support and would recommend to use tools like Joy2Key for this like you already do.
I installed it and mapped key "Z" in the tool and in the Studio.
The pause menu was working for me this way.

![image](https://github.com/syd711/vpin-studio/assets/1169826/0b2bc9d8-e004-4ef4-8675-a8d9b962a079)

**Comment by Ltek** [comment]
I cant get VPS Pause menu or overlay to work - at all... I'm using the physical keyboard and tried multiple different key set in VPS settings.

btw, Z is usually mapped to nudge in VPX, I'd suggest not using that.

**Comment by syd711** [comment]
I'll see if I can add a test button for the pause menu.
The new version will support to distribute assets on Popper screen when pause is pressed, so a preview function certainly helps to see how the screens are set up.

**Comment by syd711** [comment]
Ok, lets continue here. I added the test button for the pause menu. This should help with the setup. So the last status was that we have to figure out, why the shortcut is not working, right?

**Comment by syd711** [comment]
I added the button debounce with 2.16.0, so I assume this issue as fixed. There have been issues with the preferences refresh in the patch I gavie you. These should be fixed with the release too.

**Comment by Ltek** [comment]
@syd711 Pause menu still not working... definitely need more delay, maybe even allow it to be a user definable delay.

Test today using Joy2Key (key mapper/reader I use to map Windows DirectInout buttons to Popper stuff like 'Play only').
When I press ']' (set for VPS Overlay), VPS is sending a keyboard 'P' command immediately.

**Comment by syd711** [comment]
Mmh, damn I kept the delay hard-coded as it was in the test version I gave you (1 sec). I thought this fixed the issue. So it is still showing up and hiding again immediately? Sure, I can add a configurable flag for this. I just want to make sure before if this is really the issue. I'm suprised that it is now not working (again?).

**Comment by syd711** [comment]
I had another ticket where the workaround was to select some random key for the overlay (which was previously empty).
There is a bug that empty values are not checked properly which results in unnessessary key events.

**Comment by syd711** [comment]
If improved the key event handling and added the "debounce" option.
If the menu still does not show, please attach the latest logs here again.

**Comment by syd711** [comment]
Any updates? Otherwise I would like to close this ticket.

**Comment by Ltek** [comment]
I'll test today

**Comment by syd711** [comment]
And once again a baby(?) step forward.
2.18.5 contains a button recorder that hopefully fixes your controller issues.

**Comment by Ltek** [comment]
no go... I tried mapping 3 different keys that I had never mapped before... including 'insert' since nothing uses that. The key grabber you put it sees all the codes just fine. But when a table is running (popper in the background) the Pause menu never comes up... with button press or directly using the keyboard. I have it set to 1s debounce.

log attached
[vpin-studio-ui.log](https://github.com/syd711/vpin-studio/files/15342094/vpin-studio-ui.log)
[vpin-studio-server.log](https://github.com/syd711/vpin-studio/files/15342095/vpin-studio-server.log)

**Comment by syd711** [comment]
This issue is currently blocked by https://github.com/syd711/vpin-studio/issues/245

**Comment by syd711** [comment]
Blocking ticket has been fixed. You can try to use the VPX monitor which should now detect the game status and should not block the pause menu from opening.

**Comment by Ltek** [comment]
I'll test soon as next ver is released

**Comment by Ltek** [comment]
not ignoring this... just been working on Swamp Thing. Its almost done and I'll get back to testing VPS

**Comment by syd711** [comment]
I'm also working on a re-implementation of the key binding. The newer version will have controller support. So in combination with the VPX monitor, I'm positive that it will work much better.

**Comment by syd711** [comment]
So, we once again can continue here.
The controller support is there, so we only need to figure out now, why the table start/end commands (curl commands) are not working for you. Or maybe they are now?

**Comment by Ltek** [comment]
I tested controller mapping, its not working for me... I tried mapping both with a cabinet button (mapped to ']'), and directly with the keyboard key itself ']'.

Maybe the easiest was if to have Studio detect when 'P' key is pressed it will show Pause Menu? 'P' pauses VPX for me without any hiccups.

**Comment by syd711** [comment]
The existing controller mapping is as good as it can get right now. There is no other API I can use to detect other input methods.
So I'm running out of options here. No input is detected on the cabinet by the Studio that can be used instead?

**Comment by syd711** [comment]
Starting some "end of year" housekeeping, I'll close this ticket. As mentioned above I tried to give the best controller support I can with the latest re-implemention. Right now I'm not planning anything beyond that.

**Comment by Ltek** [comment]
This is / has been working for me... sorry I didnt report back sooner. Thank you for making this work!
